src/leap/mail/imap/service/imap.py

```diff
@@ -139,83 +108,40 @@ def buildProtocol(self, addr):
         imapProtocol.factory = self
         return imapProtocol
 
-    def doStop(self, cv=None):
+    def doStop(self):
         """
         Stops imap service (fetcher, factory and port).
-
-        :param cv: A condition variable to which we can signal when imap
-                   indeed stops.
-        :type cv: threading.Condition
-        :return: a Deferred that stops and flushes the in memory store data to
-                 disk in another thread.
-        :rtype: Deferred
         """
+        # TODO should wait for all the pending deferreds,
+        # the twisted way!
         if DO_PROFILE:
             log.msg("Stopping PROFILING")
             pr.disable()
             pr.dump_stats(PROFILE_DAT)
 
-        ServerFactory.doStop(self)
-
-        if cv is not None:
-            def _stop_imap_cb():
-                logger.debug('Stopping in memory store.')
-                self._memstore.stop_and_flush()
-                while not self._memstore.producer.is_queue_empty():
-                    logger.debug('Waiting for queue to be empty.')
-                    # TODO use a gatherResults over the new/dirty
-                    # deferred list,
-                    # as in memorystore's expunge() method.
-                    time.sleep(1)
-                # notify that service has stopped
-                logger.debug('Notifying that service has stopped.')
-                cv.acquire()
-                cv.notify()
-                cv.release()
-
-            return threads.deferToThread(_stop_imap_cb)
+        return ServerFactory.doStop(self)
 
 
 def run_service(*args, **kwargs):
     """
     Main entry point to run the service from the client.
 
-    :returns: the LoopingCall instance that will have to be stoppped
-              before shutting down the client, the port as returned by
-              the reactor when starts listening, and the factory for
-              the protocol.
+    :returns: the port as returned by the reactor when starts listening, and
+              the factory for the protocol.
     """
-    from twisted.internet import reactor
-    # it looks like qtreactor does not honor this,
-    # but other reactors should.
-    reactor.suggestThreadPoolSize(20)
-
-    leap_assert(len(args) == 2)
-    soledad, keymanager = args
+    soledad = args[0]
```

Comment:
done, i've called it "store". Although the only store in a while will be a Soledad instance, I'm more confortable having it as a generic type in case we can to change something (and that's the idea behind the adaptor).